internal/appsec/limiter_test.go

```diff
@@ -0,0 +1,305 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+//go:build appsec
+// +build appsec
+
+package appsec
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestLimiterUnit(t *testing.T) {
+	startTime := time.Now()
+
+	t.Run("no-ticks-1", func(t *testing.T) {
+		l := NewTestTicker(1, 100)
+		l.start(startTime)
+		defer l.stop()
+		// No ticks between the requests
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True")
+		require.False(t, l.Allow(), "Second call to limiter.Allow() should return False")
+	})
+
+	t.Run("no-ticks-2", func(t *testing.T) {
+		l := NewTestTicker(100, 100)
+		l.start(startTime)
+		defer l.stop()
+		// No ticks between the requests
+		for i := 0; i < 10; i++ {
+			require.True(t, l.Allow(), "Call to limiter.Allow() should return True")
+		}
+	})
+
+	t.Run("10ms-ticks", func(t *testing.T) {
+		l := NewTestTicker(1, 100)
+		l.start(startTime)
+		defer l.stop()
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True")
+		l.tick(startTime.Add(10 * time.Millisecond))
+		require.True(t, l.Allow(), "Second call to limiter.Allow() after 10ms should return True")
+	})
+
+	t.Run("9ms-ticks", func(t *testing.T) {
+		l := NewTestTicker(1, 100)
+		l.start(startTime)
+		defer l.stop()
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True")
+		l.tick(startTime.Add(9 * time.Millisecond))
+		require.False(t, l.Allow(), "Second call to limiter.Allow() after 9ms should return False")
+		l.tick(startTime.Add(18 * time.Millisecond))
+		require.True(t, l.Allow(), "Third call to limiter.Allow() after 18ms should return True")
+	})
+
+	t.Run("1s-rate", func(t *testing.T) {
+		l := NewTestTicker(1, 1)
+		l.start(startTime)
+		defer l.stop()
+		require.True(t, l.Allow(), "First call to limiter.Allow() should return True with 1s per token")
+		l.tick(startTime.Add(500 * time.Millisecond))
+		require.False(t, l.Allow(), "Second call to limiter.Allow() should return False with 1s per Token")
+		l.tick(startTime.Add(1000 * time.Millisecond))
```

Comment:
But 500 ms is enough, right?
I'd test with 500 ms instead if that's the case.

Reply:
500 *additional* ms is enough, since the rate limiter is set to 1 token/sec, so startTime + 1000ms in this case (the result of `startTime.Add` is no stored in `startTime`)